firedrake/preconditioners/offload.py

```diff
@@ -0,0 +1,117 @@
+from firedrake.preconditioners.base import PCBase
+from firedrake.functionspace import FunctionSpace, MixedFunctionSpace
+from firedrake.petsc import PETSc
+from firedrake.ufl_expr import TestFunction, TrialFunction
+import firedrake.dmhooks as dmhooks
+from firedrake.dmhooks import get_function_space
+
+__all__ = ("OffloadPC",)
+
+
+class OffloadPC(PCBase):
+    """Offload PC from CPU to GPU and back.
```

Comment:
This docstring could perhaps contain more detail about what is actually happening and even perhaps why one may wish to do this.

Reply:
E.g. This is only for CUDA GPUs